File: unleash/src/main/java/com/silvercar/unleash/event/EventDispatcher.java
package com.silvercar.unleash.event;

import com.silvercar.unleash.util.UnleashConfig;
import com.silvercar.unleash.util.UnleashScheduledExecutor;

public class EventDispatcher {

    private final UnleashSubscriber unleashSubscriber;
    private final UnleashScheduledExecutor unleashScheduledExecutor;

    public EventDispatcher(UnleashConfig unleashConfig) {
        this.unleashSubscriber = unleashConfig.getSubscriber();
        this.unleashScheduledExecutor = unleashConfig.getScheduledExecutor();
    }

    public void dispatch(UnleashEvent unleashEvent) {
        unleashScheduledExecutor.scheduleOnce(() -> {
            unleashSubscriber.on(unleashEvent);
            unleashEvent.publishTo(unleashSubscriber);
        });
    }
}

File: unleash/src/main/java/com/silvercar/unleash/event/EventDispatcher.kt
package com.silvercar.unleash.event

import com.silvercar.unleash.util.UnleashConfig
import com.silvercar.unleash.util.UnleashScheduledExecutor

class EventDispatcher(unleashConfig: UnleashConfig) {
  private val unleashSubscriber: UnleashSubscriber = unleashConfig.subscriber
  private val unleashScheduledExecutor: UnleashScheduledExecutor = unleashConfig.scheduledExecutor

  fun dispatch(unleashEvent: UnleashEvent) {
    unleashScheduledExecutor.scheduleOnce(Runnable {
      unleashSubscriber.on(unleashEvent)
      unleashEvent.publishTo(unleashSubscriber)
    })
  }
}
